### PR TITLE
Adds new property editor

### PR DESCRIPTION
**Pull request by jbrundage**
@GordonSmith @mlzummo @gururx please review

Signed-off-by: Jay Brundage <jaman.brundage@lexisnexis.com>

### CODE REVIEW

File: widgets/src/other/PropertyEditor.js [hunk 1]
```diff
@@ -98,306 +108,752 @@
...
+//                                                    })
+//                                                    ;
+//                                        })
+//                                        ;
+//                                try {
+//                                    input.node().value = JSON.stringify(widget._data);
+//                                } catch (e) {
+//                                    input.node().value = e;
+//                                }
+//                            }
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: Can this commented code be remove?

File: widgets/src/other/PropertyEditor.css [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    display:none;
+}
+.mm-label:hover {
+    cursor:pointer;
+}
+/*
+.widget-property-hover{
+    background-color: #F1C40F!important;
+    border-color:#f39c12;
+}*/
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.css]: Can this commented code be removed?

**Comment by GordonSmith** [comment]
Note:  You should really make these suggested changes on your existing branch, that way we can see the history of the pull request...

**Comment by jbrundage** [comment]
Sure, I'll remove the comments now

File: widgets/src/other/PropertyEditor.js [hunk 3]
```diff
@@ -121,276 +135,729 @@
...
-                    .enter()
-                    .append("th")
-                        .text(function (column) {
-                            return column;
-                        })
-                ;
-                th.exit()
-                    .remove()
-                ;
+        if (tableNeedsRedraw(this)) {
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: Cheating - the previous version merged changes in, the D3 way!  (but thats ok)

File: widgets/src/other/PropertyEditor.js [hunk 4]
```diff
@@ -121,276 +135,729 @@
...
+                                });
+                            });
+                        }
+                    }
+                    sPropSections.push(sectionArr);
+                }
+            }
+        }
+        //Creating Table and THEAD
+        if (true) {
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: Why?

File: widgets/src/other/PropertyEditor.js [hunk 5]
```diff
@@ -121,276 +135,729 @@
...
+            var table = element.selectAll("#" + this._id + " > table").data(this._data, function(d) {
+                return d._id;
+            });
+            table.enter().append("table")
+                    .each(function(widget) {
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: indentation is off

File: widgets/src/other/PropertyEditor.js [hunk 6]
```diff
@@ -121,276 +135,729 @@
...
+                        var element = d3.select(this);
+                        var thead = element.append("thead");
+                        if(context._collapsible_sections){
+                            thead.attr("class", "mm-label max").on("click", function() {
+                                var elm = this;
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: remove

File: widgets/src/other/PropertyEditor.js [hunk 7]
```diff
@@ -121,276 +135,729 @@
...
+                            }
+                            return text;
+                        });
+                        thead = thead.append("tr").attr("class", "mm-content");
+                        var tbody = element.append("tbody").attr("class", "mm-content");
+                        var th = thead.selectAll("th").data(context._columns, function(d) {
+                            return d;
+                        });
+                        th.enter()
+                                .append("th")
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: indentation

File: widgets/src/other/PropertyEditor.js [hunk 8]
```diff
@@ -121,276 +135,729 @@
...
+                                .text(function(column) {
+                                    return column;
+                                });
+                        th.exit().remove();
+                    });
+        }
+        //Creating TBODY
+        if (true) {
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: why?

File: widgets/src/other/PropertyEditor.js [hunk 9]
```diff
@@ -121,276 +135,729 @@
...
+                                    context.onChange(widget, "columns");
+                                })
+                        })
+                    input.node().value = JSON.stringify(widget._columns);
+                }
+                //  Data  ---
+                if (context._show_data) {
+                    var tr = tbody.append("tr");
+                    var td = tr.append("td")
+                            .text("Data")
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: indentation

File: widgets/src/other/PropertyEditor.js [hunk 10]
```diff
@@ -121,276 +135,729 @@
...
+                            ;
+                    td = tr.append("td")
+                    input = td.append("textarea")
+                            .attr("rows", "4")
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: indentation

File: widgets/src/other/PropertyEditor.js [hunk 11]
```diff
@@ -121,276 +135,729 @@
...
+                            .attr("cols", "25")
+                            .on("change", function() {
+                                widget
+                                        .data(JSON.parse(this.value))
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: indentation

File: widgets/src/other/PropertyEditor.js [hunk 12]
```diff
@@ -121,276 +135,729 @@
...
                                             .attr("class", "input_" + widget._id)
-                                            .on("change", function () {
-                                                widget[d.id](this.value).render(function (widget) {
-                                                    context.onChange(widget, d.id);
-                                                    context.render();
-                                                })
+                                            .on("change", function() {
+                                                var that = this;
+                                                d.widgetArr.forEach(function(w) {
+                                                    //
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: remove

File: widgets/src/other/PropertyEditor.js [hunk 13]
```diff
@@ -121,276 +135,729 @@
...
                                                         colorInput.on("change").apply(colorInput.node());
                                                     })
-                                                ;
-                                                inputColor.node().value = widget[d.id]();
-                                            } catch (e) {
-                                                inputColor.remove();
-                                            }
+                                                    ;
+                                            d.widgetArr.forEach(function(w) {
+                                                //
```
Comment (GordonSmith) [inline widgets/src/other/PropertyEditor.js]: remove

**Comment by GordonSmith** [comment]
Pulled via https://github.com/hpcc-systems/Visualization/pull/304

**Pull request closed without merging**